File: BeyondDEV/apps/backend/app/services/proyectos.py
from .base import BaseRepository
from .schemas.proyecto import ProyectoSchema

class ProyectosRepository(BaseRepository):
    def create_proyecto(self, data: ProyectoSchema):
        print("Proyecto a crear desde el backend: ", data)
        try:
            query = "EXEC sp_crear_proyecto %s, %s, %s, %s, %s, %s, %s, %s"
            params = (
                data.nombre,
                data.tipo,
                data.descripcion,
                data.prioridad,
                data.estado,
                data.fecha_inicio,
                data.entrega_propuesta,
                data.presupuesto
            )

            return self._execute_query(
                query,
                params,
                is_write=True,
            )
        except Exception as e:
            print(f"Error al crear el proyecto: {e}")
            return None

    def listar_proyectos(self):
        query = "EXEC sp_listar_proyectos"
        '''params = (
            data.proyecto_id
        )'''
        return self._execute_query(query)

File: BeyondDEV/apps/frontend/lib/proyectos.ts
import { useState, useEffect, useCallback } from 'react';
import { fetchAPI } from './api';
import { Use } from "./use";

export interface Proyecto {
    proyecto_id?: number;
    cliente_id?: number;
    nombre?: string;
    descripcion?: string;
    prioridad?: string;
    tipo?: string;
    fecha_inicio?: string;
    entrega_propuesta?: string;
    presupuesto?: number;
    estado?: string;
    [key: string]: any;
}

async function getProyectos(): Promise<Proyecto[]> {
    return fetchAPI<Proyecto[]>(`/proyectos`,
        {
            method: 'GET',
            params: {
                proyecto_id: null
            }
        }
    );
}

export function useProyectos() {
    const [state, setState] = useState<Use<Proyecto[]>>({
        data: null,
        loading: true,
        error: null,
    });

    const refetch = useCallback(async () => {
        setState(prev => ({ ...prev, loading: true, error: null }));
        try {
            const data = await getProyectos();
            setState({ data, loading: false, error: null });
        } catch (error) {
            setState({ data: null, loading: false, error: error as Error });
        }
    }, []);

    useEffect(() => {
        refetch();
    }, [refetch]);

    return { ...state, refetch };
}

export async function createProyecto(proyecto: Proyecto) {
    console.log("Proyecto a crear desde la API para el backend: ", proyecto);
    return fetchAPI<Proyecto>(`/proyectos`, {
        method: 'POST',
        body: JSON.stringify(proyecto),
    });
}